### PR TITLE
Bookmarking events and generic features

### PR DESCRIPTION
**Pull request by gnunicorn**
@kumarpalsinh25 as you wanted: the riverpod infrastructure (and a widget) for bookmarking.

This is what is does/looks like (right now).

https://github.com/user-attachments/assets/6eef6d31-438d-472f-80be-fd5bfe2ae1bc

Only implemented the toggle and for events only at the moment. Though the provider should make clear how to get a full list per type (`event`, `pins`, etc) and the toggle widget should work everywhere essentially.

### CODE REVIEW

File: app/lib/features/bookmarks/widgets/bookmark_action.dart [hunk 1]
```diff
@@ -0,0 +1,31 @@
+import 'package:acter/features/bookmarks/actions/bookmarking.dart';
+import 'package:acter/features/bookmarks/providers/bookmarks_provider.dart';
+import 'package:flutter/material.dart';
+import 'package:flutter_riverpod/flutter_riverpod.dart';
+
+class BookmarkAction extends ConsumerWidget {
+  final String type;
+  final String id;
```
Comment (kumarpalsinh25) [inline app/lib/features/bookmarks/widgets/bookmark_action.dart:8]:
```suggestion
  final Bookmarker bookmarker;
```

File: app/lib/features/bookmarks/widgets/bookmark_action.dart [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+  const BookmarkAction({
+    super.key,
+    required this.type,
+    required this.id,
+  });
+
+  @override
+  Widget build(BuildContext context, WidgetRef ref) {
+    final isBookmarked = ref.watch(isBookmarkedProvider((type: type, id: id)));
+
```
Comment (kumarpalsinh25) [inline app/lib/features/bookmarks/widgets/bookmark_action.dart:18]:
```suggestion
final type = bookmarker.type;
final id = bookmarker.id;
```

File: app/lib/features/bookmarks/actions/bookmarking.dart [hunk 3]
```diff
@@ -0,0 +1,20 @@
+import 'package:acter/features/bookmarks/providers/bookmarks_provider.dart';
+import 'package:flutter_riverpod/flutter_riverpod.dart';
+
+Future<bool> bookmark({
+  required WidgetRef ref,
+  required String key,
+  required id,
```
Comment (kumarpalsinh25) [inline app/lib/features/bookmarks/actions/bookmarking.dart:7]:
```suggestion
  required Bookmarker bookmarker;
```

File: app/lib/features/bookmarks/actions/bookmarking.dart [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+}) async {
+  final manager = await ref.read(bookmarksManagerProvider.future);
+  return await manager.add(key, id);
+}
+
+Future<bool> unbookmark({
+  required WidgetRef ref,
+  required String key,
+  required id,
```
Comment (kumarpalsinh25) [inline app/lib/features/bookmarks/actions/bookmarking.dart:16]:
```suggestion
  required Bookmarker bookmarker;
```

File: app/lib/features/bookmarks/actions/bookmarking.dart [hunk 5]
```diff
@@ -0,0 +1,20 @@
+import 'package:acter/features/bookmarks/providers/bookmarks_provider.dart';
+import 'package:flutter_riverpod/flutter_riverpod.dart';
+
+Future<bool> bookmark({
+  required WidgetRef ref,
+  required String key,
+  required id,
+}) async {
+  final manager = await ref.read(bookmarksManagerProvider.future);
+  return await manager.add(key, id);
```
Comment (kumarpalsinh25) [inline app/lib/features/bookmarks/actions/bookmarking.dart:10]:
```suggestion
  return await manager.add(bookmarker.type, bookmarker.id);
```

File: app/lib/features/bookmarks/actions/bookmarking.dart [hunk 6]
```diff
@@ -0,0 +1,20 @@
...
+}
+
+Future<bool> unbookmark({
+  required WidgetRef ref,
+  required String key,
+  required id,
+}) async {
+  final manager = await ref.read(bookmarksManagerProvider.future);
+  return await manager.remove(key, id);
```
Comment (kumarpalsinh25) [inline app/lib/features/bookmarks/actions/bookmarking.dart:19]:
```suggestion
  return await manager.remove(bookmarker.type, bookmarker.id);
```

File: app/lib/features/bookmarks/providers/bookmarks_provider.dart [hunk 7]
```diff
@@ -0,0 +1,22 @@
...
+import 'package:acter/features/bookmarks/types.dart';
+import 'package:acter_flutter_sdk/acter_flutter_sdk_ffi.dart';
+import 'package:flutter_riverpod/flutter_riverpod.dart';
+
+final bookmarksManagerProvider =
+    AsyncNotifierProvider<BookmarksManagerNotifier, Bookmarks>(
+  () => BookmarksManagerNotifier(),
+);
+
+final bookmarkByKeyProvider =
```
Comment (kumarpalsinh25) [inline app/lib/features/bookmarks/providers/bookmarks_provider.dart:11]:
```suggestion
final bookmarkByTypeProvider =
```

File: app/lib/features/bookmarks/providers/bookmarks_provider.dart [hunk 8]
```diff
@@ -0,0 +1,22 @@
...
+    FutureProvider.family<List<String>, String>((ref, key) async {
+  final bookmarks = await ref.watch(bookmarksManagerProvider.future);
+  return (bookmarks.entries(key)).map((s) => s.toDartString()).toList();
+});
+
+final isBookmarkedProvider =
+    StateProvider.family<bool, Bookmarker>((ref, query) {
```
Comment (kumarpalsinh25) [inline app/lib/features/bookmarks/providers/bookmarks_provider.dart:18]:
```suggestion
    StateProvider.family<bool, Bookmarker>((ref, bookmarker) {
```

File: app/lib/features/bookmarks/providers/bookmarks_provider.dart [hunk 9]
```diff
@@ -0,0 +1,22 @@
...
+  final bookmarks =
+      ref.watch(bookmarkByKeyProvider(query.type)).valueOrNull ?? [];
+  return bookmarks.contains(query.id);
```
Comment (kumarpalsinh25) [inline app/lib/features/bookmarks/providers/bookmarks_provider.dart:21]:
```suggestion
      ref.watch(bookmarkByKeyProvider(bookmarker.type)).valueOrNull ?? [];
  return bookmarks.contains(bookmarker.id);
```

**Review by kumarpalsinh25 — APPROVED** [review]
Thanks @gnunicorn for this PR. Code looks great!
I will do the needful for listing part and other module by referencing this code.

Added some minor word suggestions that doesn't stoping this PR to merge!

**Comment by gnunicorn** [comment]
@kumarpalsinh25 I've updated the code slightly to your requests but also replaced it with a typed system. I'll merge as-is but you might still want to take a second look to understand how to use this.

File: app/lib/features/bookmarks/types.dart [hunk 10]
```diff
@@ -1 +1,22 @@
-typedef Bookmarker = ({String type, String id});
+typedef Bookmarker = ({BookmarkType type, String id});
+
+// referencing the inner type
+enum BookmarkType {
```
Comment (kumarpalsinh25) [inline app/lib/features/bookmarks/types.dart:4]: Very Nice!!

**Pull request merged**